### ISSUE TITLE
[Bug]: Alignment specialization on causal_conv1d metadata pointers forces inference-time JIT compiles; raced shared-cache writes produced silent all-NaN outputs

### ISSUE THREAD

**Issue by aoshen02** [vllm-project/vllm#52413]
### Summary

`_causal_conv1d_fwd_kernel` and `_causal_conv1d_update_kernel` let Triton specialize on the 16-byte alignment of their metadata pointers (`cache_indices_ptr`, `query_start_loc_ptr`, `conv_state_indices_ptr`, ...).

Three facts make this a problem:

1. **The specialization buys nothing here.** These arrays are read one scalar at a time (a block index, an offset), so the aligned and unaligned binaries run at the same speed. Alignment specialization only pays off for wide vectorized loads (`x_ptr`, `w_ptr`, ... — not these args).
2. **Their alignment flips at runtime.** They are per-step slices of batch metadata, so the address offset depends on the batch layout of that step.
3. **Warmup can never cover the unaligned variant.** Warmup tensors are freshly allocated and therefore always aligned. So the first compile of the unaligned variant is guaranteed to happen *during inference* — this is (part of) the latency spike tracked in #43009, and `jit_monitor` (#45631) duly warns about it.

In a multi-node deployment this goes from a latency nuisance to a correctness bug.

### What we hit

GRPO RL training, Nemotron-3-Super-120B, 6 vLLM engines on 6 GB200 nodes, with a shared `TRITON_CACHE_DIR` on Lustre (for warm starts). Engines intermittently returned **all-NaN logits from the first prefill token** — only for long prompts, only on some engines, and unreproducible afterwards.

Disk forensics (no repro needed):

- These kernels compiled to disk exactly 3 times; the mtimes match the 3 NaN windows one-to-one. After the cache filled up: 40+ clean rounds, zero recurrence.
- Two cache hash dirs for the same kernel appeared **within the same minute**, differing only in the `tt.divisibility=16` attribute on `cache_indices_ptr` — two nodes concurrently compiling the aligned/unaligned variants into the same shared dir. Triton's cache group write (`__grp__*.json` + cubin) is not atomic across writers.
- Every failing request was ≥143K tokens (its `cache_indices` slice lands at an unaligned offset); every success ≤138K. Only some engines were poisoned — the race losers.

A race loser ends up running the aligned-assumption binary on unaligned `cache_indices` → reads wrong block indices → garbage → all-NaN logits, silently.

Sharing `TRITON_CACHE_DIR` across nodes is on us, and we now scope it per node. But the variant axis itself has no upside, and single-node users still pay the inference-time compile spike for it.

### Proposed fix

Add the metadata pointers to `do_not_specialize_on_alignment` — the same mechanism #50000 already applied to `num_cache_lines` in these exact kernels:

```python
@triton.jit(do_not_specialize_on_alignment=[
    "num_cache_lines",
    "cache_indices_ptr", "has_initial_states_ptr", "query_start_loc_ptr",
    "batch_ptr", "token_chunk_offset_ptr", "block_idx_first_scheduled_token",
    "block_idx_last_scheduled_token", "initial_state_idx", "num_computed_tokens",
])
def _causal_conv1d_fwd_kernel(...)

@triton.jit(do_not_specialize_on_alignment=[
    "num_cache_lines", "batch",
    "conv_state_indices_ptr", "num_accepted_tokens_ptr", "query_start_loc_ptr",
    "block_idx_last_scheduled_token", "initial_state_idx",
])
def _causal_conv1d_update_kernel(...)
```

Effect: one binary per config, compiled without the alignment assumption, correct for any address. Warmup covers it naturally, nothing left to JIT during inference, no variants to race. Zero perf cost (scalar loads only); the wide data pointers keep their specialization.

Relation to #48363: complementary. That PR warms the fwd kernel by sweeping 8 alignment combinations; with the specialization removed the sweep collapses to a single call, and `_causal_conv1d_update_kernel` (not covered there) is fixed as well.

We're happy to open a PR (change verified against Triton 3.6: the flag drops the divisibility bit from the specialization key for pointer args, `triton/runtime/jit.py`).

### Environment

Observed on vLLM 0.25.1 (aarch64, GB200, Triton 3.6.0); the kernel decorators are unchanged on current `main` (f97e502969).

**Comment by harshillodhiya** [comment]
Hey team! 

I've submitted this fix for the critical race condition in causal_conv1d (issue #52413). 

**The TL;DR:** Triton was specializing on metadata pointer alignment per-request, causing JIT recompiles and cache write races. The fix is clean—just tell Triton not to specialize on these pointers, and we get a stable kernel with zero data corruption risk.

I've added a regression test that validates the fix and prevents future regressions. It passes locally, and I'm ready to run the full suite once you give it a look.

Would appreciate any feedback on the approach or tests! Thanks for reviewing.

**Comment by aoshen02** [comment]
> Hey team!
> 
> I've submitted this fix for the critical race condition in causal_conv1d (issue [#52413](https://github.com/vllm-project/vllm/issues/52413)).
> 
> **The TL;DR:** Triton was specializing on metadata pointer alignment per-request, causing JIT recompiles and cache write races. The fix is clean—just tell Triton not to specialize on these pointers, and we get a stable kernel with zero data corruption risk.
> 
> I've added a regression test that validates the fix and prevents future regressions. It passes locally, and I'm ready to run the full suite once you give it a look.
> 
> Would appreciate any feedback on the approach or tests! Thanks for reviewing.

Hi, still investigating the real cause, will just close it right now.
